File: src/Csharp/packages/base/src/Exceptions.cs
using System.IO;
using System.Runtime.Hosting;

/*! 
    \brief Custom exceptions thrown by DHARTAPI Internally
    
    \remarks
    Exceptions are created here as needed, and generally existing C# excepions 
    are preferred.
*/
namespace DHARTAPI.Exceptions
{
	/*! 
     \brief Common error codes to convey the state of the program.
     \remarks Used as the return type for most native functions.
    */
	internal enum HF_STATUS
	{
		/// <summary>
		/// The equested operation executed successfully.
		/// </summary>
		OK = 1,

		// Special Codes //
		/// <summary>
		/// This functionality has not yet been implemented. This should not be seen at all in release.
		/// </summary>
		NOT_IMPLEMENTED = -54,

		// Error Codes //
		/// <summary>
		/// A generic error code often used as a catch all. Instances of this code being returned
		/// should be made more specific using another error code.
		/// </summary>
		GENERIC_ERROR = 0,

		/// <summary>
		/// The file at the given path was not found.
		/// </summary>
		NOT_FOUND = -1,

		/// <summary>
		/// The provided mesh was invalid.
		/// </summary>
		INVALID_MESH = -2,

		/// <summary>
		/// No graph was produced by the given input.
		/// </summary>
		NO_GRAPH = -3,

		/// <summary>
		/// Unused.
		/// </summary>
		INVALID_COST = -4,

		/// <summary>
		/// An external dependency required for this function to execute was not found.
		/// </summary>
		MISSING_DEPEND = -5,

		/// <summary>
		/// Ran out of memory during execution, usually because the dataset was too big.
		/// </summary>
		OUT_OF_MEMORY = -6,

		/// <summary>
		/// Unused.
		/// </summary>
		MALFORMED_DB = -7,

		/// <summary>
		/// Unused.
		/// </summary>
		DB_BUSY = -8,

		/// <summary>
		/// One or more of the given pointers was invalid.
		/// </summary>
		INVALID_PTR = -9,

		/// <summary>
		/// The requested element was out of the range of the container.
		/// </summary>
		OUT_OF_RANGE = -10,

		/// <summary>
		/// No path could be found between the start and endpoints given. Used exclusively for pathfinding.
		/// </summary>
		NO_PATH = -11,

		/*! \brief There is no cost with the given name associated with the graph */
		NO_COST = -12,

		/*! \brief The graph has not been compressed! */
		NOT_COMPRESSED = -13


	};

	/*! \brief The file passed was not an obj file, or was an invalid obj file */
	public class InvalidMeshException : System.ApplicationException
	{
		/// <summary>Initializes a new instance of the <see cref="InvalidMeshException" /> class.</summary>
		/// <param name="message">The message that describes the error.</param>
		public InvalidMeshException(string message) : base(message) { }
	}

	/*! \brief Tried to add an alternate cost to an edge without meeting the preconditions. */
	public class InvalidCostOperation : System.ApplicationException
	{
		public InvalidCostOperation(string message) : base(message) { }
	}

	/*! \brief There was some logical error with what the user was trying to perform. */
	public class LogicError : System.ApplicationException
	{
		public LogicError(string message) : base(message) { }
	}
}

File: src/Python/dhart/Exceptions.py
from enum import IntEnum as Enum


class HF_STATUS(Enum):
    """ Error codes to be returned from C++ """

    OK = 1
    NOT_IMPLEMENTED = -54
    GENERIC_ERROR = 0  # Not sure what happened here (If this gets thrown, either fix it or give it a status code!)
    NOT_FOUND = -1  # A file does not exist at the given path
    INVALID_OBJ = -2  # The given path did not point to a valid obj file
    NO_GRAPH = -3  # This requires a valid graph in the DB to execute successfully
    INVALID_COST = -4  # The given cost name does not exist in the database
    MISSING_DEPEND = -5  # A dependency for this object is missing.
    OUT_OF_MEMORY = -6  # Ran out of memory during the last operation
    MALFORMED_DB = -7  # The database exists, but is in some kind of error state
    DB_BUSY = -8  # The database is busy (is there some external connection?)
    EMPTY_INPUT = -9  # Nothing was passed!
    OUT_OF_RANGE = (
        -10
    )  # The requested datapoint was out of the bounds of the given structure
    NO_PATH = -11  # No path was returned from the pathfinding ooperation
    NO_COST = -12  # The given cost could not be found.
    NOT_COMPRESSED = -13  # The graph needed to be compressed and it wasn't


class HFException(Exception):
    """ Generic DHARTAPIException """
    def __init__(self, message):
        super().__init__(message)

class InvalidOBJException(HFException):
    """ The provided An invalid OBJ file was specified.

    This can be raised if the given filepath leads to a file that is not an
    OBJ file, the file is a malformed obj file.
    """
    def __init__(self, message = ""):
        super().__init__(message)


class FileNotFoundException(HFException):
    """ The file at the given path given file was not found """
    def __init__(self, message = ""):
        super().__init__(message)


class OutOfRangeException(HFException):
    """ The requested datapoint was out of the bounds of the given structure """
    def __init__(self, message = ""):
        super().__init__(message)


class MissingDependencyException(HFException):
    """ One or more of the required dependencies for this operation were unable to be found """
    def __init__(self, message = ""):
        super().__init__(message)

class LogicError(HFException):
    """ A class was used improperly. """
    def __init__(self, message = ""):
        super().__init__(message)

class InvalidCostOperation(HFException):
    """ Tried to add an edge to a cost that didn't already exist in
        the default cost set 
    """
    def __init__(self, message = ""):
        super().__init__(message)